dp3t-sdk/sdk/src/main/java/org/dpppt/android/sdk/DP3T.java

```diff
@@ -246,71 +247,110 @@ public static void checkGaenAvailability(Context context, Consumer<GaenAvailabil
 
 	public static void sendIAmInfected(Activity activity, Date onset, ExposeeAuthMethod exposeeAuthMethod,
 			ResponseCallback<DayDate> callback) {
+
+		showShareTEKsPopup(activity, new ResponseCallback<Void>() {
+
+			@Override
+			public void onSuccess(Void response) {
+				uploadTEKs(activity, onset, exposeeAuthMethod, callback);
+			}
+
+			@Override
+			public void onError(Throwable throwable) {
+				reportFailedIAmInfected(throwable, callback);
+			}
+		});
+	}
+
+	public static void showShareTEKsPopup(Activity activity, ResponseCallback<Void> callback) {
+		checkInit();
+		pendingShareTEKsPopup = new PendingShareTEKsPopup(callback);
+		showShareTEKsPopupInternal(activity);
+	}
+
+	public static void uploadTEKs(Context context, Date onset, ExposeeAuthMethod exposeeAuthMethod,
+			ResponseCallback<DayDate> callback) {
 		checkInit();
+		if (pendingShareTEKsPopup == null || pendingShareTEKsPopup.temporaryExposureKeys == null) {
+			throw new IllegalStateException("showShareTEKsPopup() must be successfully executed before calling uploadTEKs()");
+		}
+		uploadTEKsInternal(context, pendingShareTEKsPopup.temporaryExposureKeys, onset, exposeeAuthMethod, callback);
+	}
 
-		pendingIAmInfectedRequest = new PendingIAmInfectedRequest(onset, exposeeAuthMethod, callback);
+	private static void showShareTEKsPopupInternal(Activity activity) {
+		if (pendingShareTEKsPopup == null) {
+			throw new IllegalStateException("pendingShareTEKsPopup must be set before calling showShareTEKsPopup()");
+		}
 
-		executeIAmInfected(activity);
+		GoogleExposureClient.getInstance(activity).getTemporaryExposureKeyHistory(activity, REQUEST_CODE_EXPORT_KEYS,
+				temporaryExposureKeys -> {
+					pendingShareTEKsPopup.temporaryExposureKeys = temporaryExposureKeys;
+					pendingShareTEKsPopup.callback.onSuccess(null);
```

Comment:
`pendingShareTEKsPopup` should be set to `null` to avoid memory leaks.

Reply:
Not here though, but in `uploadTEKs()` after using the data. (Although would the retry work in case of an upload error?)

Reply:
I think we might use an object/callback and return that the caller of those methods instead of using a static object pendingShareTEKsPopup. That would avoid leaks in case of an error, or if the flow gets aborted.